agent/immortalstreams/backedpipe/backed_reader_test.go

```diff
@@ -0,0 +1,612 @@
+	data     []byte
+	pos      int
+	err      error
+	readFunc func([]byte) (int, error)
+}
+
+func newMockReader(data string) *mockReader {
+	return &mockReader{data: []byte(data)}
+}
+
+func (mr *mockReader) Read(p []byte) (int, error) {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+
+	if mr.readFunc != nil {
+		return mr.readFunc(p)
+	}
+
+	if mr.err != nil {
+		return 0, mr.err
+	}
+
+	if mr.pos >= len(mr.data) {
+		return 0, io.EOF
+	}
+
+	n := copy(p, mr.data[mr.pos:])
+	mr.pos += n
+	return n, nil
+}
+
+func (mr *mockReader) setError(err error) {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+	mr.err = err
+}
+
+func TestBackedReader_NewBackedReader(t *testing.T) {
+	t.Parallel()
+
+	errorChan := make(chan backedpipe.ErrorEvent, 1)
+	br := backedpipe.NewBackedReader(errorChan)
+	require.NotNil(t, br)
+	require.Equal(t, uint64(0), br.SequenceNum())
+	require.False(t, br.Connected())
+}
+
+func TestBackedReader_BasicReadOperation(t *testing.T) {
+	t.Parallel()
+	ctx := testutil.Context(t, testutil.WaitShort)
+
+	errorChan := make(chan backedpipe.ErrorEvent, 1)
+	br := backedpipe.NewBackedReader(errorChan)
+	reader := newMockReader("hello world")
+
+	// Connect the reader
+	seqNum := make(chan uint64, 1)
+	newR := make(chan io.Reader, 1)
+
+	go br.Reconnect(seqNum, newR)
+
+	// Get sequence number from reader
+	seq := testutil.RequireReceive(ctx, t, seqNum)
+	require.Equal(t, uint64(0), seq)
+
+	// Send new reader
+	testutil.RequireSend(ctx, t, newR, io.Reader(reader))
+
+	// Read data
+	buf := make([]byte, 5)
+	n, err := br.Read(buf)
+	require.NoError(t, err)
+	require.Equal(t, 5, n)
+	require.Equal(t, "hello", string(buf))
+	require.Equal(t, uint64(5), br.SequenceNum())
+
+	// Read more data
+	n, err = br.Read(buf)
+	require.NoError(t, err)
+	require.Equal(t, 5, n)
+	require.Equal(t, " worl", string(buf))
+	require.Equal(t, uint64(10), br.SequenceNum())
+}
+
+func TestBackedReader_ReadBlocksWhenDisconnected(t *testing.T) {
+	t.Parallel()
+	ctx := testutil.Context(t, testutil.WaitShort)
+
+	errorChan := make(chan backedpipe.ErrorEvent, 1)
+	br := backedpipe.NewBackedReader(errorChan)
+
+	// Start a read operation that should block
+	readStarted := make(chan struct{}, 1)
+	readDone := make(chan struct{})
+	var readErr error
+	var readBuf []byte
+	var readN int
+
+	go func() {
+		defer close(readDone)
+		// Signal that we're about to start the read
+		readStarted <- struct{}{}
+		buf := make([]byte, 10)
+		readN, readErr = br.Read(buf)
+		readBuf = buf[:readN]
+	}()
+
+	// Wait for read to start
+	testutil.RequireReceive(ctx, t, readStarted)
+
+	// Ensure the read is actually blocked by verifying it hasn't completed
+	// and that the reader is still not connected
+	require.Eventually(t, func() bool {
+		select {
+		case <-readDone:
+			t.Fatal("Read should be blocked when disconnected")
+			return false
+		default:
+			// Read is still blocked, which is what we want
+			return !br.Connected()
+		}
+	}, testutil.WaitShort, testutil.IntervalMedium)
```

Comment:
I don't understand why this is wrapped in `Eventually` --- the reader has never connected at this point, so `br.Connected() == true` should be a test failure, not a retry. The other branch of the select is also a test failure.